src/components/tab-bar/tabs.ts

```diff
@@ -0,0 +1,47 @@
+import { Tab } from './tab.types';
+
+/**
+ * Set a tabs `active` state to true in a list of tabs. The previous tab with
+ * `active` set to true will have it set to `false` instead.
+ *
+ * @param {Tab[]} tabs list of tabs
+ * @param {number} index the index of the tab to set to active
+ *
+ * @returns {Tab[]} a copy of the list of tabs with the changed tabs replaced
+ */
+export function setActiveTab(tabs: Tab[], index: number): Tab[] {
+    const oldSelectedTab = tabs.find((t) => t.active === true);
+    let result = tabs;
+
+    if (oldSelectedTab) {
+        const deselectTab = getMapCompatibleReplaceTabFunction({
+            ...oldSelectedTab,
+            active: false,
+        });
+        result = result.map(deselectTab);
+    }
+
+    const selectTab = getMapCompatibleReplaceTabFunction({
+        ...result[index],
+        active: true,
+    });
+
+    return result.map(selectTab);
+}
+
+function getMapCompatibleReplaceTabFunction(updatedTab: Tab) {
+    return replaceTab;
+
+    // eslint-disable-next-line prefer-arrow/prefer-arrow-functions
+    function replaceTab(existingTab: Tab) {
```

Comment:
The lint rule wants this to like as follows:

```ts
function getMapCompatibleReplaceTabFunction(updatedTab: Tab) {
    const replaceTab = (existingTab: Tab) => {
        if (updatedTab.id === existingTab.id) {
            // This is the tab we want to update, so return the
            // updated tab instead of the existing one.
            return updatedTab;
        }

        // This is not the tab we are looking for,
        // so just return it untouched.
        return existingTab;
    };

    return replaceTab;
}
```

But I think that makes it less readable, because it lures me into reading the details of the inner function first, and trying to juggle more details in my mind at once.

Reply:
Hmm I'm not sure about this refactor, to me this is less readable :slightly_frowning_face: 
But, reading your refactor makes me think I know what is causing the confusion for you :slightly_smiling_face: I'm guessing that it is the curried mapping function? We could remove that whole function probably and use `splice` in the `setActiveTab` function instead. I normally don't like `splice` since it mutates the array, but if we make a copy of it first it should be fine.

So something like this maybe?

```ts
export function setActiveTab(tabs: Tab[], index: number): Tab[] {
    const oldSelectedIndex = tabs.findIndex((t) => t.active === true);    
    let result = [...tabs];

    if (oldSelectedIndex) {
        result.splice(oldSelectedIndex, 1, { ...tabs[oldSelectedIndex], active: false });
    }

    result.splice(index, 1, { ...tabs[index], active: true });

    return result;
}
```

Reply:
`splice` should have a performance benefit as well, O(1) vs O(n) 
Although it's probably very minor since the number of items will be quite small :sweat_smile:

Reply:
And we don't actually have to use `splice` lol
We could just use
```ts
result[index] = { ...tabs[index], active: true }
```
:upside_down_face:

Reply:
The curried function itself isn't really a problem for me, but the combination of the arrow-syntax and the initial call of the function inside the map() made it a stumbling block. Then the use of `oldSelectedTab` and `oldTab` to mean *completely* different things came in and kicked me when I was down 😆

Not to toot my own horn here, but I think I'm relatively well acquainted with higher order programming in javascript, so if I'm having trouble following the code, I think there's a good chance other people might too.

Regarding the splice-idea, I hadn't thought of using `splice` specifically, but I did think about whether we could maybe just find the relevant elements and flip their `active` value, and then use spread to make a copy of the array to return, if that's needed to trigger any re-rendering?